presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,560 @@
+            extends PushdownDereferencesInProject<ProjectNode>
+    {
+        PushDownDereferenceThroughProject()
+        {
+            super(project());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, ProjectNode projectNode, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            ImmutableMap<VariableReferenceExpression, RowExpression>
+                    dereferencesMap =
+                    expressions.inverse().entrySet().stream().collect(toImmutableMap(Map.Entry::getKey,
+                            entry -> castToRowExpression(entry.getValue())));
+
+            return Result.ofPlanNode(
+                    new ProjectNode(context.getIdAllocator().getNextId(),
+                            projectNode.getSource(),
+                            Assignments.builder().putAll(projectNode.getAssignments()).putAll(dereferencesMap).build()));
+        }
+    }
+
+    public class PushDownDereferenceThroughUnnest
+            extends PushdownDereferencesInProject<UnnestNode>
+    {
+        PushDownDereferenceThroughUnnest()
+        {
+            super(unnest());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, UnnestNode unnestNode, BiMap<DereferenceExpression, VariableReferenceExpression> expressions)
+        {
+            // Create new Project contains all pushdown symbols above original source
+            ImmutableMap<VariableReferenceExpression, RowExpression>
+                    dereferencesMap =
+                    expressions.inverse().entrySet().stream().collect(toImmutableMap(Map.Entry::getKey,
+                            entry -> castToRowExpression(entry.getValue())));
+            Assignments assignments = Assignments.builder().putAll(identityAssignmentsAsSymbolReferences(unnestNode.getSource().getOutputVariables())).putAll(dereferencesMap).build();
+            ProjectNode source = new ProjectNode(context.getIdAllocator().getNextId(), unnestNode.getSource(), assignments);
+
+            // Create new UnnestNode
+            UnnestNode newUnnest = new UnnestNode(context.getIdAllocator().getNextId(),
+                    source,
+                    ImmutableList.<VariableReferenceExpression>builder().addAll(unnestNode.getReplicateVariables()).addAll(expressions.values()).build(),
+                    unnestNode.getUnnestVariables(),
+                    unnestNode.getOrdinalityVariable());
+            return Result.ofPlanNode(newUnnest);
+        }
+    }
+
+    private static PlanNode createProjectBelow(PlanNode planNode, Assignments dereferences, PlanNodeIdAllocator idAllocator)
+    {
+        if (dereferences.isEmpty()) {
+            return planNode;
+        }
+        return new ProjectNode(idAllocator.getNextId(), planNode, Assignments.builder().putAll(identityAssignmentsAsSymbolReferences(planNode.getOutputVariables())).putAll(dereferences).build());
```

Comment:
perhaps, extract assignments variable for readability:

```
        Assignments assignments = Assignments.builder()
                .putAll(identityAssignmentsAsSymbolReferences(planNode.getOutputVariables()))
                .putAll(dereferences)
                .build();
        return new ProjectNode(idAllocator.getNextId(), planNode, assignments);
```

Reply:
fixed